records/2023-06-27.scalability.increasing.send.rates.md

```diff
@@ -0,0 +1,107 @@
+# Scalability: increasing send rates
+
+Date: 2023-06-27
+
+## Status
+
+**DRAFT**.
+
+## Context
+
+As Notify usage grows (and in particular as we anticipate increasing usage by PTMs) we need to increase our email and SMS send rates.
+
+## Initial Goals
+
+We should be able to approach the rates that AWS sets.
+
+### Email
+
+Currently in staging and production our email limit with SES is [100 emails / second or 200K / day](https://ca-central-1.console.aws.amazon.com/ses/home?region=ca-central-1#/account). This can be raised on request. 
+
+### SMS
+
+Currently we are using [long codes in PinPoint](https://ca-central-1.console.aws.amazon.com/pinpoint/home?region=ca-central-1#/sms-account-settings/phoneNumbers) to send SMS. Each long code can send one SMS fragment / second. On staging we have one long code and on production we have 20. So our AWS limits are 1 fragment / second on staging and 20 fragments / second on production. We can request more long codes, though there is a limit (TODO: what is the limit?). Note that the limit for each short code is 100 fragments / second, so switching to short codes would substantially increase our AWS SMS send limit.
+
+## What happens when we send a lot of emails?
+
+Emails are batch saved into the database. This does not appear to be a bottleneck. Each email is send by its own `deliver_email` task.
+
+[`deliver_email()`](https://github.com/cds-snc/notification-api/blob/main/app/celery/provider_tasks.py#L63)
+- Fetches the notification using the id
+- Calls `send_email_to_provider()`
+- So basically a wrapper for send_email_to_provider()
+
+[`send_email_to_provider()`](https://github.com/cds-snc/notification-api/blob/main/app/delivery/send_to_providers.py#L191)
+- Prepares the email
+- Get the template
+- Fill it out
+- Get any files if they’re attached
+- Calls `send_email()` with the html and any attachments
+
+[`send_email()`](https://github.com/cds-snc/notification-api/blob/main/app/clients/email/aws_ses.py#L33)
+- Creates a MIME email with the html and attachments
+- Calls boto3 `send_raw_email()` with the MIME email
+
+So there are basically 3 areas that can take time
+- `send_email_to_provider()` code before the `send_email()` call
+- `send_email()` code before the boto call
+- the boto call
+
+We have timing metrics for:
+- `deliver_email()` (which should be the same as `send_email_to_provider()`): celery_tasks_deliver_email
+- `send_email()`: celery_clients_ses_send_email
+- `send_raw_email()`:  celery_clients_ses_request-time
+
+
+We uploaded a csv of 50K emails (with one variable and no file attachments) to staging (send to `success@simulator.amazonses.com`) on July 12, 2:15 pm EST and monitored the [scaling dashboard](https://ca-central-1.console.aws.amazon.com/cloudwatch/home?region=ca-central-1#dashboards/dashboard/ScalingTemp?start=2023-07-12T18%3A10%3A00Z&end=2023-07-12T19%3A30%3A00Z&autoRefresh=false).
```

Comment:
When we send a batch of 50k emails like this, do we do 50k lookups to get the template, one for each notification that goes out? 🤔

Reply:
yeah... there's probably some caching we could use that would have an impact on large sends